.github/workflows/deploy_bookdown.yml

```diff
@@ -0,0 +1,52 @@
+on:
+  push:
+     branches:
+       - master
+
+  
+
+name: renderbook
+
+jobs:
+  bookdown:
+    name: Render-Book
+    runs-on: macOS-latest
+    steps:
+      - uses: actions/checkout@v1
+      - uses: r-lib/actions/setup-r@v1
+      - uses: r-lib/actions/setup-pandoc@v1
+      - name: Install rmarkdown
+        run: Rscript -e 'install.packages(c("rmarkdown","bookdown", "car", "afex", "psych", "factoextra"))'
+      - name: Render Book
+        run: Rscript -e 'bookdown::render_book("index.Rmd")'
+      - uses: actions/upload-artifact@v1
+        with:
+          name: _book
+          path: _book/
+  
+# Need to first create an empty gh-pages branch
+# see https://pkgdown.r-lib.org/reference/deploy_site_github.html
+# and also add secrets for a GH_PAT and EMAIL to the repository
+# gh-action from Cecilapp/GitHub-Pages-deploy
+  checkout-and-deploy:
```

Comment:
Why do we need two steps?

Basically, you're doing:

1. Step one
  1. render some stuff
  2. upload it
2. Step two
  1. Download results from step one
  2. publish them online

It would make more sense to do:

1. Step one
  1. render some stuff
  2. upload the results (if we still need to do this)
  3. publish the results online

Also, the bandwidth for `actions/upload-artifact` and `actions/upload-artifact` is not unlimited, so we should take care not to run this too often.

Reply:
So in the documentation of the workflow (https://orchid00.github.io/actions_sandbox/websites-using-pkgdown-bookdown-and-blogdown.html) it states that:
"This action is performed using two jobs, the first renders html, and the second deploys the html to the gh-pages branch so that it can be viewed using the url <user/org_name>.github.io/<repository_name>. It achieves this by passing an artifact (the _book directory) between jobs."

So it might have to do with the results being uploaded to a different branch? This would be a clean way to only have the html files in the branch from which the github page is rendered, but I am not sure if it is strictly necessary. It would probably also work to render them to the same branch (or maybe a folder on that branch?)

Reply:
well, that link explains what the action does, but not *why* it is doing it that way. Did you enable the action on your own fork to test it? That way you can just try it out and see what works/ if the two steps are necessary.

Reply:
I did try it, yes. And it does it as I described. We could probably rewrite the script such that it does not create an artifact that is then passed to the other branch, but saves it to the current branch, if that is preferable?

Reply:
Something like this: https://github.community/t/how-to-make-a-workflow-file-to-save-the-output-as-a-file-to-my-github-repo/18352/2

Reply:
> We could probably rewrite the script such that it does not create an artifact that is then passed to the other branch, but saves it to the current branch, if that is preferable?

Well, I still don't know why we need to save this anywhere. My idea would be something like this:
```yaml
on:
  push:
     branches:
       - master

name: renderbook

jobs:
  bookdown:
    name: Render-Book
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v1
      - uses: r-lib/actions/setup-r@v1
      - uses: r-lib/actions/setup-pandoc@v1
      - name: Install rmarkdown
        run: Rscript -e 'install.packages(c("rmarkdown","bookdown", "car", "afex", "psych", "factoextra"))'
      - name: Render Book
        run: Rscript -e 'bookdown::render_book("index.Rmd")'

     - name: Deploy to GitHub Pages
       uses: Cecilapp/GitHub-Pages-deploy@3.2.1
       env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
       with:
          build_dir: _book/ # "_site/" by default
          email: ${{ secrets.EMAIL }}  # must be a verified email
```

Reply:
From my understanding, you need to save the html files in a repo to generate a Github page from it. I am not sure if this is possible without repo. What would the url of the github page then be? As it is usually the address of the repo? 

I have tried out the code you suggested but I am running into errors. When I use Ubuntu, it fails to render the book:
![image](https://user-images.githubusercontent.com/61797391/152968826-d60afbb3-2f7f-4aba-939b-3bdc2d4f3360.png)
This is the error message: 
![image](https://user-images.githubusercontent.com/61797391/152968911-b6dae7aa-934e-42ed-a380-b5b3761cc329.png)

When I run it on MacOS, the book renders, but the Deploy to GitHub Pages does not work:
![image](https://user-images.githubusercontent.com/61797391/152969111-7686dba0-150a-4b27-a0d4-e378e2e4c3bc.png)

Reply:
Can you post a link to the full log files? I wasn't saying that the code I wrote would work straight out of the box, but just that that is essentially what the code should do (so without the uploading and downloading steps).

Uploading to the gh-pages branch can also be done using something like (but probably not exactly this!):
```
git config --local user.email "actions@github.com"
git config --local user.name "GitHub Actions"
git add .
git commit -m "update book"
git push origin gh-pages
```
That's essentially what pkgdown does.

Reply:
Sure, see the last 2 workflows: https://github.com/JTPetter/jasp-vp-test/actions.

Right, then we would not need the Cecilapp/GitHub-Pages-deploy@3.2.1 at all, I believe. Makes me wonder what it does. Because the html files are created in the previous step.

Reply:
I have tried to rewrite the actions script, such that it pushes the html files directly to the repo. But I keep getting an error message. 
`on:
  push:
     branches:
       - master

name: renderbook

jobs:
  bookdown:
    name: Render-Book
    runs-on: macOS-latest

    steps:
      - uses: actions/checkout@v1
      - uses: r-lib/actions/setup-r@v1
      - uses: r-lib/actions/setup-pandoc@v1

      - name: Install rmarkdown
        run: Rscript -e 'install.packages(c("rmarkdown","bookdown", "car", "afex", "psych", "factoextra"))'

      - name: Render Book
        run: Rscript -e 'bookdown::render_book("index.Rmd")'
        
      - name: Config
        run: |
          git config --local user.email "actions@github.com"
          git config --local user.name "GitHub Actions"
        
      - name: Push to GitHub Pages branch
        run: |
          git add .
          git commit -m "update book"
          git push origin master`

and the error is:

`error: src refspec master does not match any`

Do you have an idea what might be going on?

Reply:
@vandenman

Reply:
It could be that the brain called main and not master? But don't you want to push to gh-pages?

Reply:
Ok so I tried a few things and I think that this should do what we want:

https://github.com/vandenman/jasp-vp-test/blob/master/.github/workflows/deploy_bookdown.yml

Reply:
Hi Don,
thanks for looking into this! Your code works fine. I had to switch the OS to Mac, however. When I use Linux, the workflow fails in rendering the book. Looking at the error message and timing, it seems that it crashes as soon as it has to read one of the .csv files. Do you have any idea why that might be? This is the error message:
![image](https://user-images.githubusercontent.com/61797391/156935142-0155683a-8b52-4a2b-b32d-33dc47c7ae78.png)

Reply:
Yeah that's another thing you should probably fix, see my changes here: https://github.com/JTPetter/jasp-vp-test/compare/master...vandenman:master

Whenever you refer to filenames, please make sure they use the exact case as the file on disk. This doesn't matter on Windows or macOS, so `read.csv("myFile.csv")` happily loads a file called `myfile.csv` but on Ubuntu (or any other Linux flavor) this does not work.

See also https://stat.ethz.ch/R-manual/R-devel/library/base/html/files.html and the section Case-insensitive file systems.

Reply:
Ah, good to know! I got it to work now and have pushed the updated workflow to this PR. Thanks for looking into this!